add Day 17 article (#87)
scalacenter/scala-advent-of-code: docs/puzzles/day17.md

https://adventofcode.com/2021/day/17

## Solution of Part 1

### Modelling The Domain

#### A Moving Probe

The problem asks us to consider the trajectory of a projectile _probe_
which has both a _position_ and a _velocity_. Both positions and velocities
have two directions, `x` and `y`, with integer values. We will model them
with case classes:

```scala
case class Position(x: Int, y: Int)
```
```scala
case class Velocity(x: Int, y: Int)
```

and then the probe is as follows:

```scala
case class Probe(position: Position, velocity: Velocity)
```

We also find out that a probe always has an initial position of `0,0`, so
we will model that also:

```scala
val initial = Position(x = 0, y = 0)
```

We are also told about the projectile motion of the probe - it moves
in discrete steps, which we will implement as follows:
- on each step we create a new probe with a new position and a new velocity,
- the `x` of the new position is the sum of `x` of the old position and `x`
  of the old velocity,
- the `y` of the new position is the sum of `y` of the old position and `y`
  of the old velocity,
- the `x` of the new velocity is given by subtracting the sign (`-1`, `0`

>   or `1`) of the `x` of the old velocity from itself (drag slows the

Reviewer: ```suggestion
  or `1`) of the `x` of the old velocity from itself (drag slows down the
```

  projectile),
- the `y` of the new velocity is given by subtracting `1` from the `y` of
  the old velocity (due to gravity).

The code is written as such:
```scala
def step(probe: Probe): Probe =
  val Probe(pos, vel) = probe
  Probe(
    Position(x = pos.x + vel.x, y = pos.y + vel.y),
    Velocity(x = vel.x - vel.x.sign, y = vel.y - 1)
  )
```

#### A Target Area

Next we are told that a successful launch trajectory will cause the probe
to be within in a target area after at least one of its steps. The area is
defined by the points within a given range in the `x` and `y` directions.
We can model a target area by a case class with two ranges:
```scala
case class Target(xs: Range, ys: Range)
```

### Reasoning about the Problem

We are told to first identify the initial velocity of trajectories that will
hit the target area exactly, meaning that after a given step, the probe will
be exactly within the target area; and that we must disregard trajectories

> that "overshoot" i.e. pass through the entire target area in a single step.

Reviewer: ```suggestion
that "overshoot", i.e., that pass through the entire target area in a single step.
```

We are also told to find the maximum height (`y` position) reached out of
all valid trajectories.

#### Simulating a trajectory

> For this problem we will simulate a probe moving along a trajectory, i.e.

Reviewer: ```suggestion
For this problem we will simulate a probe moving along a trajectory, i.e.,
```

situations for each direction:
- for the `x` direction:
  - the `x` velocity is `0` and the `x` position of the probe is less than
    the minimum `x` position of the target,
  - the `x` position of the probe is greater than the maximum `x` position
    of the target.
- for the `y` direction:
  - the `y` velocity is less than `0` and the
    `y` position of the probe is less than the minimum `y` position of the
    target.

The code to compute this is given as such:
```scala
def beyond(probe: Probe, target: Target): Boolean =
  val Probe(pos, vel) = probe
  val Target(xs, ys) = target
  val beyondX = (vel.x == 0 && pos.x < xs.min) || pos.x > xs.max
  val beyondY = vel.y < 0 && pos.y < ys.min
  beyondX || beyondY
```

:::info
The above conditions make the assumptions that the `x` velocity is never
negative, and that the target is always in the positive `x` direction.
They are also informed by the fact that the probe will eventually always
have negative velocity (due to gravity).
:::

##### Running the Simulation

We can use our two conditions to now simulate the trajectory of a probe:
- We begin with an initial `probe`, and an initial maximum height `maxY` of
  `0`,

> - We will then iterate these values - on each iteration, we will apply

Reviewer: ```suggestion
- We then iterate these values - on each iteration, we apply
```
Avoid the future tense here. There are more "will"'s down below, which should also be removed.

  `step` to the probe, and replace `maxY` by the maximum of `maxY` and
  the `y` position of the current probe,

> - we will ignore any iteration steps where the `probe` does not collide

Reviewer: ```suggestion
- we will ignore any iteration step where the `probe` does not collide
```

>   with, or go beyond the target, i.e. the probe is still on a valid

Reviewer: ```suggestion
  with, or go beyond the target, i.e., the probe is still on a valid
```

> We then call `headOption` to get an `Option` wrapping the first element
> we are interested in. Finally `collect` allows us to inspect `probe`
> and `maxY`, and keep `maxY` when `probe` matches the condition we want,
> otherwise if the condition is not met `None` will be returned.

#### Checking all Possible Trajectories

So far we have seen how to simulate the trajectory of a single
probe. We need to find the best possible height reached by all
trajectories - meaning that we will need to generate some initial
velocities for the probe.

We can use some knowledge to help us reduce the search space for
possible velocities.

First, we assume that the target will always be in a positive direction
from the probe's initial direction, and that the probe's `x` velocity will
only get closer to `0`, so we will not need to consider
negative `x` velocities.

Second, we know that the problem requires us to find the highest positive
height reached by the probe, and that the probe's `y` velocity can only
fall once in motion. So we will not need to consider negative `y`
velocities.

That gives us the lower bounds for `x` and `y` velocities, what about

> the upper bound for the velocities?

Reviewer: ```suggestion
the upper bounds for the velocities?
```

We know that a trajectory is invalid if it goes beyond the target in a
single step, so the largest single step that the probe can move is the
distance of the furthest corner of the target. For our problem that is the

> lower right corner, so our upper bounds will be the position of that corner.

Reviewer: This explanation lacks the critical detail that, for *positive* initial y velocities, eventually there will be a point where the position is on the x axis (y position is 0) with a y velocity of `-initialYVelocity - 1`. This is why we can use `-bottomRightY` as upper bound for the y velocity.

Author: I have now addressed this comment

To proceed we create a function `allMaxHeights` to return a sequence
of possible maximum heights, one for each valid initial velocity. It
runs the simulation on each possible velocity within the bounds we
defined:

```scala
def allMaxHeights(target: Target): Seq[Int] =
  val Target(xs, ys) = target
  val upperBoundX = xs.max
  val upperBoundY = ys.min.abs
  for
    vx <- 0 to upperBoundX
    vy <- 0 to upperBoundY
    maxy <- simulate(Probe(initial, Velocity(vx, vy)), target)
  yield
    maxy
```

### Computing the Solution

#### Parsing the Input

The input for this problem is a single line, possibly ending in
a new line char. e.g.
```scala
"target area: x=20..30, y=-10..-5\n"
```

From this input we will extract two `Range` values by pattern matching.

Values of type `PartialFunction[A, B]` can be used as extractors in
pattern matching, and as we are parsing strings, let's make a type alias
`Parser[A]` to communicate our intent:
```scala
type Parser[A] = PartialFunction[String, A]
```

> First, lets make a parser for `Int` values. We use a regex to check

Reviewer: ```suggestion
First, let us make a parser for `Int` values. We use a regex to check
```

```scala
case class Probe(position: Position, velocity: Velocity)
```

We also find out that a probe always has an initial position of `0,0`, so
we model that also:

```scala
val initial = Position(x = 0, y = 0)
```

We are also told about the projectile motion of the probe - it moves
in discrete steps, which we implement as follows:
- on each step we create a new probe with a new position and a new velocity,
- the `x` of the new position is the sum of `x` of the old position and `x`
  of the old velocity,
- the `y` of the new position is the sum of `y` of the old position and `y`
  of the old velocity,
- the `x` of the new velocity is given by subtracting the sign (`-1`, `0`
  or `1`) of the `x` of the old velocity from itself (drag slows down the
  projectile),
- the `y` of the new velocity is given by subtracting `1` from the `y` of
  the old velocity (due to gravity).

The code is written as such:
```scala
def step(probe: Probe): Probe =
  val Probe(pos, vel) = probe
  Probe(
    Position(x = pos.x + vel.x, y = pos.y + vel.y),
    Velocity(x = vel.x - vel.x.sign, y = vel.y - 1)
  )
```

#### A Target Area

Next we are told that a successful launch trajectory causes the probe

> to be within in a target area after at least one of its steps. The area is

Reviewer 2: Is this really correct to say “within in”?

Reviewer: Ah no, it should be "within a target", indeed. ;)

Author: good spot, I fixed it
